### PR TITLE
Enact policy after three failed elections, fix way how the tracker advances

### PR DESCRIPTION
**Pull request by Terverok**
Adds lacking implementation of enacting a new card after three failed elections.

### CODE REVIEW

File: events/SocketEvents.js [hunk 1]
```diff
@@ -167,23 +167,33 @@ module.exports = function (io, RoomsManager) {
...
+                    if (threeVotesFailed) {
+			const topCard = RoomsManager.getTopPolicy(socket.currentRoom)
+			RoomsManager.enactPolicy(socket.currentRoom, topCard)
+			io.sockets.in(socket.currentRoom).emit(SocketEvents.NewPolicy, {
+			    data: {
+				policy: topCard,
+				timestamp: getCurrentTimestamp(),
+			    },
+			})
+		    }
```
Comment (Dammic) [inline events/SocketEvents.js]: There is one space missing here

File: src/features/Notifications/Notifications.jsx [hunk 2]
```diff
@@ -13,13 +13,16 @@ export class Notifications extends React.PureComponent {
         notifications: PropTypes.arrayOf(PropTypes.object),
         notificationsActions: PropTypes.objectOf(PropTypes.func),
     }
+
+    deleteNotificationRightWay = id => this.props.notificationsActions.deleteNotification({ id })
```
Comment (Dammic) [inline src/features/Notifications/Notifications.jsx]: You can keep this the way it was before, and in notificationDuck do something like that:
```javascript
[deleteNotification]: (state, action) => {                                                                                                                                                                                                                            
             const { notifications } = state                                                                                                                                                                                                                                   
             const id = action.payload  
...
```

File: events/SocketEvents.js [hunk 3]
```diff
@@ -167,23 +167,33 @@ module.exports = function (io, RoomsManager) {
...
-                        remaining: RoomsManager.getRemainingVotesCount(socket.currentRoom),
+                        remaining: 0,
                         timestamp: getCurrentTimestamp(),
                     },
                 })
                 io.sockets.in(socket.currentRoom).emit(SocketEvents.VOTING_PHASE_REVEAL, {
                     data: {
                         votes: RoomsManager.getVotes(socket.currentRoom),
                         timestamp: getCurrentTimestamp(),
+                        failedElections: RoomsManager.getFailedElections(socket.currentRoom),
```
Comment (Dammic) [inline events/SocketEvents.js]: You should rename that variable to ex 'failedElectionsCount'

File: src/features/PlayerBoard/PlayerBoard.jsx [hunk 4]
```diff
@@ -91,7 +91,7 @@ export class PlayerBoard extends React.PureComponent {
             }
         })
 
-        if (trackerPosition >= 3) this.props.roomActions.resetTracker()
+        if (trackerPosition > 3) this.props.roomActions.resetTracker()
```
Comment (Dammic) [inline src/features/PlayerBoard/PlayerBoard.jsx]: On trackerPosition === 3 we need to reset the tracker.

**Comment by Dammic** [comment]
When we reject max amount of possible rejections and then all vote 'yes', the tracker does not reset on backend (the next reject will result in enacted random policy)

File: utils/RoomsManager.js [hunk 5]
```diff
@@ -1,10 +1,10 @@
 const {
-    reject, findIndex, sortBy, values, tail, countBy, mapValues, isNil,
-    filter, includes, forEach, random, slice, times, map,
+    reject, findIndex, sortBy, values, tail, countBy, mapValues, isNil, isEmpty,
+    filter, includes, forEach, random, slice, times, map, head,
     find, pick, shuffle, size, sample, get, concat, fill, take, drop, pullAt, indexOf,
 } = require('lodash')
 const { GamePhases, PlayerRole, PlayerAffilications, PolicyCards } = require('../Dictionary')
-
+//go to [policies]
```
Comment (Dammic) [inline utils/RoomsManager.js]: This comment has no value to any programmer except for you

File: utils/RoomsManager.js [hunk 6]
```diff
@@ -368,7 +376,16 @@ class RoomsManager {
     getDrawnCards(roomName) {
         return this.rooms_props[roomName].drawnCards
     }
-
+    getTopPolicy(roomName) {
```
Comment (Dammic) [inline utils/RoomsManager.js]: You could merge that function with getChoicePolicyCards - just pass the amount of cards to take.
The function can also be renamed as 'takeChoicePolicyCards' as it modifies data.

Comment (Terverok) [reply utils/RoomsManager.js]: So you agree it was a good idea to rename :)

File: events/SocketEvents.js (hunk 1 above)
Comment (Terverok) [reply events/SocketEvents.js]: Where? I couldn't find it, everything seems fine to me.

File: src/features/Notifications/Notifications.jsx (hunk 2 above)
Comment (Terverok) [reply src/features/Notifications/Notifications.jsx]: done

File: events/SocketEvents.js (hunk 3 above)
Comment (Terverok) [reply events/SocketEvents.js]: done

File: src/features/PlayerBoard/PlayerBoard.jsx (hunk 4 above)
Comment (Terverok) [reply src/features/PlayerBoard/PlayerBoard.jsx]: Set to >= 3 and moved to componentDidUpdate method

File: utils/RoomsManager.js [hunk 7]
```diff
@@ -352,6 +356,10 @@ class RoomsManager {
             player.isDead = true
         }
     }
+    /*********************************************/
+    /**************policies***********************/
```
[the remark this answers is not in the document]
Comment (Terverok) [reply utils/RoomsManager.js]: done

File: utils/RoomsManager.js (hunk 6 above)
Comment (Terverok) [reply utils/RoomsManager.js]: done

File: events/SocketEvents.js [hunk 8]
```diff
@@ -8,6 +8,13 @@ module.exports = function (io, RoomsManager) {
         sendError: (socket, errorMessage) => {
             socket.emit(SocketEvents.CLIENT_ERROR, { error: errorMessage })
         },
+        sendAllChatMessage: (socket, chatMessage) => {
```
Comment (Dammic) [inline events/SocketEvents.js]: sendAllChatMessage -> broadcastChatMessage

File: events/SocketEvents.js [hunk 9]
```diff
@@ -193,13 +201,21 @@ module.exports = function (io, RoomsManager) {
...
                             : null
                         ),
                     },
                 })
+                if (!hasVotingSucceed || threeVotesFailed) {
+                    socketEvents.sendAllChatMessage(socket, 'Next turn will begin in 3 seconds!')
+                }
+                if (hasVotingSucceed || threeVotesFailed) {
+                    io.sockets.in(socket.currentRoom).emit(SocketEvents.ResetTracker, {
+                        timestamp: getCurrentTimestamp(),
```
Comment (Dammic) [inline events/SocketEvents.js]: There is no data object here

File: events/SocketEvents.js (hunk 8 above)
Comment (Dammic) [reply events/SocketEvents.js]: Anyway, you can just call sendMessage(socket, {
 content: chatMessage,
}) here

File: events/SocketEvents.js [hunk 10]
```diff
@@ -319,6 +336,7 @@ module.exports = function (io, RoomsManager) {
                 setTimeout(() => {
                     socketEvents.startChancellorChoicePhase(socket)
                 }, 3000)
+                    socketEvents.sendAllChatMessage(socket, 'Next turn will begin in 3 seconds!')
```
Comment (Dammic) [inline events/SocketEvents.js]: Too many spaces here

File: src/features/PlayerBoard/PlayerBoard.jsx [hunk 11]
```diff
@@ -36,6 +36,10 @@ export class PlayerBoard extends React.PureComponent {
         }
     }
 
+    componentDidUpdate() {
```
Comment (Dammic) [inline src/features/PlayerBoard/PlayerBoard.jsx]: I would personally put it into componentWillReceiveProps and then also check:
```javascript
 if (this.props.trackerPosition >= 3 && this.props.trackerPosition !== nextProps.trackerPosition) {
    delay(this.props.roomActions.resetTracker, 1000)
}
```

File: src/utils/SocketHandler.js [hunk 12]
```diff
@@ -140,6 +138,16 @@ export class SocketHandler extends React.PureComponent {
             this.props.notificationsActions.addNotification({ type: MessagesTypes.ERROR,  message: error })
         })
 
+        socket.on(SocketEvents.ResetTracker, (payload) => {
+            const { timestamp } = payload
```
Comment (Dammic) [inline src/utils/SocketHandler.js]: payload.data if you change it on server

File: utils/RoomsManager.js [hunk 13]
```diff
@@ -45,6 +45,8 @@ class RoomsManager {
     setChancellor(roomName) {
         const { playersDict, chancellorCandidateName } = this.rooms_props[roomName]
 
+        this.rooms_props[roomName].failedElections = 0
```
Comment (Dammic) [inline utils/RoomsManager.js]: failedElectionsCount

Comment (Terverok) [reply utils/RoomsManager.js]: done

File: src/utils/SocketHandler.js (hunk 12 above)
Comment (Terverok) [reply src/utils/SocketHandler.js]: done

File: events/SocketEvents.js (hunk 10 above)
Comment (Terverok) [reply events/SocketEvents.js]: done

File: events/SocketEvents.js (hunk 9 above)
Comment (Terverok) [reply events/SocketEvents.js]: done

File: events/SocketEvents.js (hunk 8 above)
Comment (Terverok) [reply events/SocketEvents.js]: done

File: src/features/PlayerBoard/PlayerBoard.jsx (hunk 11 above)
Comment (Terverok) [reply src/features/PlayerBoard/PlayerBoard.jsx]: corrected in componentDidUpdate, i want do it after the render

**Pull request merged**